Allow use of proxy server by publisher and gateway machines

mcb30 opened the pull request:
This patch series adds the ability to specify a proxy server to be used by a publisher or gateway machine in order to download files from the Stratum 0 repository.

This is particularly useful when the Stratum 0 files are held in cloud storage such as AWS S3: it allows for the use of an authenticating proxy (to avoid exposing the S3 reader credentials directly to every client) and opens up the possibility of using the proxy to cache files uploaded by a publisher so that the gateway does not need to waste time immediately re-downloading those same files from the cloud storage.

mcb30 committed:
Use "Cache-Control: max-age=0" instead of bypassing system proxy

The get_item() shell function has an optional "noproxy" parameter.
The purpose of this parameter is unclear, but the most plausible
explanation seems to be as a means of ensuring that the download
fetches the current upstream file, ignoring any versions that may have
previously been cached in intermediate proxy servers.

This fails on any system where use of a proxy server is required for
all downloads, since direct origin server fetches are not permitted on
such systems.

Unconditionally add the HTTP/1.1 standard request header
"Cache-Control: max-age=0" to force an origin server download (or
revalidation), thereby eliminating the need to bypass the system
proxy.

Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

mcb30 committed:
Allow use of a proxy server during repository creation

Allow for situations in which a proxy server must be used to download
files during the creation of a Stratum 0 or Stratum 1 repository.

Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

mcb30 committed:
Allow use of a proxy server by cvmfs_swissknife

Allow for cvmfs_swissknife to be used in situations in which a proxy
is required in order to access the Stratum 0 repository.

Use the '-@' option character since all alphabetic option characters
are already used by at least one cvmfs_swissknife subcommand, and
cvmfs_swissknife does not support long options.

Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

mcb30 committed:
Allow use of a proxy server for uploads to S3 storage

Allow a proxy to be specified via CVMFS_S3_PROXY in the s3.conf file.
This can be useful when the proxy is shared between several publishers
and a gateway machine since it allows the proxy to be configured to
cache uploaded files, thereby eliminating an extra round-trip when the
gateway needs to retrieve files that were uploaded by the publisher.

Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

mcb30 committed:
Allow use of a proxy server by cvmfs_publish

Allow for cvmfs_publish to be used in situations in which a proxy is
required to access the Stratum 0 repository, using a proxy server URL
configured via CVMFS_SERVER_PROXY in the repository server.conf file.

Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

DrDaveD commented on cvmfs/server/cvmfs_server_common.sh:
```diff
@@ -503,21 +503,15 @@ is_due_auto_garbage_collection() {
-    curl -f $(get_x509_cert_settings) $(get_follow_http_redirects_flag) \
-      "$url" 2>/dev/null | tr -d '\0'
-  fi
+  curl -f -H "Cache-Control: no-cache" $(get_curl_proxy) \
```
I recommend using `Cache-Control: max-age=0` instead of `no-cache`.  They're close, but not exactly the same thing.  With `max-age=0` a proxy may use a cached copy if it has been proven to be valid with `If-Modified-Since`, where `no-cache` means it absolutely cannot reuse a cached copy.  `max-age=0` is gentler on caches.  See the user agent section of this [stackover flow answer](https://stackoverflow.com/a/1383359/10457761) for more details.

mcb30 replied:
Sure, that's an easy change and doesn't materially impact the code.  I'll update it now.

mcb30 replied:
Updated to use `max-age=0` instead

jblomer commented:
@cernvm-bot all

jblomer commented:
@cernvm-bot help

jblomer commented:
@cernvm-bot cpplint

jblomer commented:
@cernvm-bot tidy

jblomer commented:
@cernvm-bot unittest

jblomer commented:
@cernvm-bot cloudtest

jblomer commented on cvmfs/publish/settings.h:
```diff
@@ -357,18 +357,21 @@ class SettingsRepository {
     : fqrn_(fqrn)
     , url_(std::string("http://localhost/cvmfs/") + fqrn_())
     , tmp_dir_("/tmp")
+    , proxy_("")
```
Let's change the order of the members and put `url_` and `proxy_` together

mcb30 replied:
Done

jblomer commented on cvmfs/publish/settings.cc:
```diff
@@ -456,6 +468,8 @@ SettingsRepository SettingsBuilder::CreateSettingsRepository(
     settings.SetTmpDir(arg + "/tmp");
   if (options_mgr_->GetValue("X509_CERT_BUNDLE", &arg))
     settings.SetCertBundle(arg);
+  if (options_mgr_->GetValue("CVMFS_HTTP_PROXY", &arg))
```
We should as well set the proxy in `CreateSettingsPublisherFromSession()` from the `user.proxy` extended attribute.

mcb30 replied:
Done

jblomer commented on cvmfs/receiver/params.h:
```diff
@@ -34,6 +34,7 @@ struct Params {
   size_t max_weight;
   size_t min_weight;
   bool upload_stats_db;
+  std::string proxy;
```
Also here let's put `stratum0` and `proxy` together.

mcb30 replied:
Done

jblomer commented:
We are just getting used to clang-tidy checks. The warnings are not related to the patch, afaics. I'll open a separate PR to fix the code surrounding the changes.

jblomer commented on cvmfs/server/cvmfs_server_common.sh:
```diff
@@ -556,6 +550,19 @@ get_x509_cert_settings() {
 
+# Parse proxy server for curl command
+get_curl_proxy() {
+  if [ x"$CVMFS_HTTP_PROXY" != x"DIRECT" ]; then
```
I think we should also ignore an empty environment variable

mcb30 replied:
After switching to `CVMFS_SERVER_PROXY`, the test is now:
```
if [ x"$CVMFS_SERVER_PROXY" != x"" ]; then
```
since `CVMFS_SERVER_PROXY` will be defined only if it contains a proxy URL (i.e. it will never exist with the value `DIRECT` anyway)

jblomer commented on cvmfs/server_tool.cc:
```diff
@@ -34,6 +35,11 @@ bool ServerTool::InitDownloadManager(const bool follow_redirects,
   download_manager_->SetRetryParameters(kDownloadRetries, 2000, 5000);
   download_manager_->UseSystemCertificatePath();
 
+  if (proxy != "") {
```
With the possibility to explicitly set a proxy server, I think we should remove the "use_system_proxy" makeshift mechanism.

mcb30 replied:
I've added a separate commit to remove the `use_system_proxy` mechanism.

jblomer reviewed (changes requested):
Many thanks!  I think an explicit proxy setting for the server side is quite useful!

For certain stratum 1 sites, we already had the possibility of using the system http proxy (`http_proxy` environment variable).  I think we should drop that in favor of administrators explicitly setting a proxy.

It might make sense to use a different parameter name, e.g. `CVMFS_SERVER_HTTP_PROXY`, because on the server side we only support a single proxy and not the client-side load-balancing configuration.

We would also need an integration test that creates a repository and makes a small publish (we can later extend the test if we need to). The test could for instance use Apache's built-in proxy, or setup a squid on the fly (or another forward proxy), whatever is simplest.

mcb30 commented:
Thanks for the feedback!  I'll get on this now and have a new version ready within a couple of hours

mcb30 commented:
> It might make sense to use a different parameter name, e.g. `CVMFS_SERVER_HTTP_PROXY`, because on the server side we only support a single proxy and not the client-side load-balancing configuration.

Good idea.  I've switched to `CVMFS_SERVER_PROXY` in the `server.conf` file.

mcb30 committed:
Remove use_system_proxy mechanism

Now that CVMFS supports the use of an explicit server-side proxy (via
CVMFS_SERVER_PROXY in the server.conf file), remove the old mechanism
allowing for the use of the system-wide proxy server.

Requested-by: Jakob Blomer <jblomer@cern.ch>
Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

mcb30 commented:
@jblomer All comments should be addressed apart from the integration test, which I'm putting together now.  Please feel free to review the updated commits, since I'm not planning any further code changes at this point.

mcb30 committed:
Add integration test for creating a repository via a proxy

Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

mcb30 commented:
I've added a basic integration test.  Looking at the cloudtest failures now...

jblomer commented:
@cernvm-bot cloudtest

mcb30 commented:
@jblomer I can't immediately see any reason for the consistent cloudtest failures across x86_64 (but passing on x86).  Does the cloudtest environment utilise any kind of system proxy server?

mcb30 commented:
Found the problem - a missing `$` in one of the scripts.  Please abort the current cloudtest run

jblomer commented:
@mcb30 They are aborted.

mcb30 commented:
I've force-pushed the fix - should be worth restarting the tests now, thanks.

jblomer commented:
@cernvm-bot cloudtest

mcb30 commented:
> cloudtests finished: FAILURE
> https://lcgapp-services.cern.ch/cvmfs-jenkins/job/CvmfsCloudTesting/1181/

The few remaining cloudtest failures now all appear to be unrelated to this PR.

jblomer commented:
Could you rebase to the latest devel branch? I meanwhile tried to address the clang-tidy errors in the surroundings of the PR.

jblomer commented:
@cernvm-bot unittest

jblomer commented:
@cernvm-bot cpplint

jblomer commented:
@cernvm-bot tidy

mcb30 commented:
Rebased onto latest devel and updated to fix the linter warning

jblomer commented:
@cernvm-bot tidy

jblomer commented:
@cernvm-bot cpplint
